bigquery/start.py

```diff
@@ -0,0 +1,224 @@
+import os.path
+import requests, json
+
+from google.auth.transport.requests import Request
+from google.oauth2.credentials import Credentials
+from google_auth_oauthlib.flow import InstalledAppFlow
+from googleapiclient.discovery import build
+from googleapiclient.errors import HttpError
+
+# If modifying these scopes, delete the file token.json.
+SCOPES = ["https://www.googleapis.com/auth/bigquery"]
+
+WEATHER_API_KEY = "YOUR_KEY"
+
+
+def create_dataset(creds, project_id, dataset_id):
+    try:
+        service = build("bigquery", "v2", credentials=creds)
+        service.datasets().insert(
+            projectId=project_id,
+            body={
+                "datasetReference": {
+                    "projectId": project_id,
+                    "datasetId": dataset_id,
+                }
+            },
+        ).execute()
+    except HttpError as err:
+        print(err)
+
+
+def create_table(creds, project_id, dataset_id, table_id, table_schema):
+    try:
+        service = build("bigquery", "v2", credentials=creds)
+        service.tables().insert(
+            projectId=project_id,
+            datasetId=dataset_id,
+            body={
+                "tableReference": {
+                    "projectId": project_id,
+                    "datasetId": dataset_id,
+                    "tableId": table_id,
+                },
+                "schema": {"fields": table_schema},
+            },
+        ).execute()
+    except HttpError as err:
+        print(err)
+
+
+def create_city_geo(creds, project_id, dataset_id, table_id, city):
+    try:
+        service = build("bigquery", "v2", credentials=creds)
+        res = requests.get(
+            f"http://api.openweathermap.org/geo/1.0/direct?q={city},California,US&limit=5&appid={WEATHER_API_KEY}"
+        )
+        res = json.loads(res.text)
+        for data in res:
+            if "local_names" in data:
+                del data["local_names"]
+            if data["lon"] > -121.8:
+                continue
+            service.tabledata().insertAll(
+                projectId=project_id,
+                datasetId=dataset_id,
+                tableId=table_id,
+                body={"rows": [{"json": data}]},
+            ).execute()
+
+    except HttpError as err:
+        print(err)
+
+
+def main():
```

Comment:
If I understand correctly, this is a setup script. Unless I'm mistaken, one pitfall with it is that it only needs to be run once -and yet we should be clear about the consequences of it being run again.

Reply:
After I remove the geo function, the rest of the code could be run again. It will just return some error without having any negative effect on the running dataset.